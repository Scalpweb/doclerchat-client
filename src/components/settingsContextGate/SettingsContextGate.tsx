import React, {useState, useEffect, useRef, useContext} from 'react';
import {TranslationContext} from 'components/translationContextGate/TranslationContextGate';

export interface Settings {
  username: string;
  interfaceColor: 'light' | 'dark';
  clockDisplay: '12' | '24';
  enableControlEnter: 'on' | 'off';
  language: string;
}

export const defaultSettings: Settings = {
  username: 'John Doe',
  interfaceColor: 'light',
  clockDisplay: '12',
  enableControlEnter: 'off',
  language: 'en'
};

export interface SettingsContextProps {
  fetch: () => Settings;
  store: (settings: Settings) => void;
  storeOne: (key: string, value: string | boolean) => void;
}

export const SettingsContext = React.createContext<SettingsContextProps>({
  fetch: () => defaultSettings,
  store: () => {
    return;
  },
  storeOne: () => {
    return;
  }
});

const STORAGE_KEY = 'doclerChatSettings';

interface Props {
  children: (settings: Settings) => React.ReactNode;
}

const fetchSettings = () => {
  const savedSettings = window.localStorage.getItem(STORAGE_KEY);
  if (savedSettings) {
    return JSON.parse(savedSettings);
  }
  return defaultSettings;
};

const SettingsContextGate = ({children}: Props) => {
  const translation = useRef(useContext(TranslationContext));
  const [settings, setSettings] = useState<Settings>(fetchSettings());

  const storeSettings = (newSettings: Settings) => {
    window.localStorage.setItem(STORAGE_KEY, JSON.stringify(newSettings));
    setSettings({...newSettings});
  };

  const storeOneSettings = (key: string, value: string) => {
    storeSettings({...fetchSettings(), [key]: value});
  };

  useEffect(() => {
    translation.current.setLanguage(fetchSettings()['language']);
  }, []);

  return (
    <SettingsContext.Provider
      value={{fetch: fetchSettings, store: storeSettings, storeOne: storeOneSettings}}>
      {children(settings)}
    </SettingsContext.Provider>
  );
};

export default SettingsContextGate;
