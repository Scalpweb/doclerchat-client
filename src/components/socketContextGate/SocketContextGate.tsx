import React, {useEffect, useState, useRef, useContext} from 'react';
import {SettingsContext} from 'components/settingsContextGate/SettingsContextGate';
import SocketManager, {Messages} from '../../utils/socket';
import Loader from 'components/loader/Loader';
import {makeTabBlink} from '../../utils/utils';

export interface SocketContextProps {
  send: (type: string, json: string) => void;
}

export const SocketContext = React.createContext<SocketContextProps>({
  send: () => {
    return;
  }
});

interface Props {
  children: (messages: Messages[], unread: number, resetUnreadCount: () => void) => React.ReactNode;
}

const SocketContextGate = ({children}: Props) => {
  const context = useRef(useContext(SettingsContext));

  const [connected, setConnected] = useState(false);
  const [messages, setMessages] = useState<Messages[]>([]);
  const [unreadCount, setUnreadCount] = useState(0);

  const update = () => {
    context.current.storeOne('username', SocketManager.getInstance().getUsername());
    setMessages([...SocketManager.getInstance().getMessages()]);
    setConnected(SocketManager.getInstance().isConnected());
    setUnreadCount((u) => u + 1);
    makeTabBlink();
  };

  const sendMessage = (type: string, data: string) => {
    SocketManager.getInstance().send(type, data);
  };

  useEffect(() => {
    SocketManager.getInstance().init(context.current.fetch().username);
    SocketManager.getInstance().addListener(update);
    update();
    return () => {
      SocketManager.getInstance().removeListener(update);
    };
  }, []);

  const resetUnreadCount = () => {
    setUnreadCount(0);
  };

  if (!connected) {
    return <Loader />;
  }

  return (
    <SocketContext.Provider value={{send: sendMessage}}>
      {children(messages, unreadCount, resetUnreadCount)}
    </SocketContext.Provider>
  );
};

export default SocketContextGate;
