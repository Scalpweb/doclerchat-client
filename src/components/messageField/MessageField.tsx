import React, {useState, useRef, useContext} from 'react';
import Picker, {IEmojiData} from 'emoji-picker-react';
import {SocketContext} from 'components/socketContextGate/SocketContextGate';
import IconButton from '../iconButton/IconButton';
import {SettingsContext} from 'components/settingsContextGate/SettingsContextGate';

const MessageField = () => {
  const socket = useRef(useContext(SocketContext));
  const settings = useRef(useContext(SettingsContext));

  const [message, setMessage] = useState('');
  const [showPicker, setShowPicker] = useState(false);

  const sendMessage = () => {
    socket.current.send('write', message);
    setMessage('');
  };

  const handleKeyUp = (e: React.KeyboardEvent<HTMLTextAreaElement>) => {
    if (e.ctrlKey || e.metaKey) {
      if (e.keyCode === 13 && settings.current.fetch().enableControlEnter === 'on') {
        sendMessage();
      }
    }
  };

  const pickEmoji = (_: Event, emoji: IEmojiData) => {
    setMessage((m) => m + emoji.emoji);
    setShowPicker(false);
  };

  return (
    <>
      {showPicker && <Picker onEmojiClick={pickEmoji} />}
      <textarea
        value={message}
        onKeyUp={handleKeyUp}
        onChange={(e) => setMessage(e.currentTarget.value)}
      />
      <div className="pull-left">
        <IconButton icon="insert_emoticon" onClick={() => setShowPicker((s) => !s)} />
      </div>
      <div className="pull-right">
        <IconButton icon="send" onClick={sendMessage} />
      </div>
    </>
  );
};

export default MessageField;
