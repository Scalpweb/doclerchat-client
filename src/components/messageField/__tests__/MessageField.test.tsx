import React from 'react';
import {mount} from 'enzyme';
import {SocketContext} from '../../socketContextGate/SocketContextGate';
import {SettingsContext} from '../../settingsContextGate/SettingsContextGate';
import MessageField from '../MessageField';

jest.mock('emoji-picker-react', () => {
  return {
    __esModule: true,
    // eslint-disable-next-line react/display-name
    default: () => {
      return <div></div>;
    }
  };
});

describe('MessageField', () => {
  it('renders emoji selector', () => {
    const component = mount(
      <SettingsContext.Provider value={{fetch: jest.fn(), store: jest.fn(), storeOne: jest.fn()}}>
        <SocketContext.Provider value={{send: jest.fn()}}>
          <MessageField />
        </SocketContext.Provider>
      </SettingsContext.Provider>
    );
    expect(component).toMatchSnapshot();

    component.find('.pull-left IconButton').simulate('click');
    expect(component).toMatchSnapshot();
  });

  it('clears textarea when sending message', () => {
    const component = mount(
      <SettingsContext.Provider value={{fetch: jest.fn(), store: jest.fn(), storeOne: jest.fn()}}>
        <SocketContext.Provider value={{send: jest.fn()}}>
          <MessageField />
        </SocketContext.Provider>
      </SettingsContext.Provider>
    );
    component.find('textarea').getDOMNode<HTMLTextAreaElement>().value = 'foobarbaz';
    component.find('textarea').simulate('change');
    expect(component).toMatchSnapshot();

    component.find('.pull-right IconButton').simulate('click');
    expect(component).toMatchSnapshot();
  });
});
