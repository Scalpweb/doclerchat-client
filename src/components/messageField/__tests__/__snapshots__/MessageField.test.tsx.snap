// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageField clears textarea when sending message 1`] = `
<MessageField>
  <textarea
    onChange={[Function]}
    onKeyUp={[Function]}
    value="foobarbaz"
  />
  <div
    className="pull-left"
  >
    <IconButton
      icon="insert_emoticon"
      onClick={[Function]}
    >
      <a
        href="#"
        onClick={[Function]}
      >
        <span
          className="material-icons"
        >
          insert_emoticon
        </span>
      </a>
    </IconButton>
  </div>
  <div
    className="pull-right"
  >
    <IconButton
      icon="send"
      onClick={[Function]}
    >
      <a
        href="#"
        onClick={[Function]}
      >
        <span
          className="material-icons"
        >
          send
        </span>
      </a>
    </IconButton>
  </div>
</MessageField>
`;

exports[`MessageField clears textarea when sending message 2`] = `
<MessageField>
  <textarea
    onChange={[Function]}
    onKeyUp={[Function]}
    value=""
  />
  <div
    className="pull-left"
  >
    <IconButton
      icon="insert_emoticon"
      onClick={[Function]}
    >
      <a
        href="#"
        onClick={[Function]}
      >
        <span
          className="material-icons"
        >
          insert_emoticon
        </span>
      </a>
    </IconButton>
  </div>
  <div
    className="pull-right"
  >
    <IconButton
      icon="send"
      onClick={[Function]}
    >
      <a
        href="#"
        onClick={[Function]}
      >
        <span
          className="material-icons"
        >
          send
        </span>
      </a>
    </IconButton>
  </div>
</MessageField>
`;

exports[`MessageField renders emoji selector 1`] = `
<MessageField>
  <textarea
    onChange={[Function]}
    onKeyUp={[Function]}
    value=""
  />
  <div
    className="pull-left"
  >
    <IconButton
      icon="insert_emoticon"
      onClick={[Function]}
    >
      <a
        href="#"
        onClick={[Function]}
      >
        <span
          className="material-icons"
        >
          insert_emoticon
        </span>
      </a>
    </IconButton>
  </div>
  <div
    className="pull-right"
  >
    <IconButton
      icon="send"
      onClick={[Function]}
    >
      <a
        href="#"
        onClick={[Function]}
      >
        <span
          className="material-icons"
        >
          send
        </span>
      </a>
    </IconButton>
  </div>
</MessageField>
`;

exports[`MessageField renders emoji selector 2`] = `
<MessageField>
  <default
    onEmojiClick={[Function]}
  >
    <div />
  </default>
  <textarea
    onChange={[Function]}
    onKeyUp={[Function]}
    value=""
  />
  <div
    className="pull-left"
  >
    <IconButton
      icon="insert_emoticon"
      onClick={[Function]}
    >
      <a
        href="#"
        onClick={[Function]}
      >
        <span
          className="material-icons"
        >
          insert_emoticon
        </span>
      </a>
    </IconButton>
  </div>
  <div
    className="pull-right"
  >
    <IconButton
      icon="send"
      onClick={[Function]}
    >
      <a
        href="#"
        onClick={[Function]}
      >
        <span
          className="material-icons"
        >
          send
        </span>
      </a>
    </IconButton>
  </div>
</MessageField>
`;
